server/src/main/java/io/druid/server/SegmentManager.java

```diff
@@ -148,49 +206,52 @@ private Segment getAdapter(final DataSegment segment) throws SegmentLoadingExcep
 
   public void dropSegment(final DataSegment segment) throws SegmentLoadingException
   {
-    String dataSource = segment.getDataSource();
-    synchronized (lock) {
-      VersionedIntervalTimeline<String, ReferenceCountingSegment> loadedIntervals = dataSources.get(dataSource);
+    final String dataSource = segment.getDataSource();
 
-      if (loadedIntervals == null) {
-        log.info("Told to delete a queryable for a dataSource[%s] that doesn't exist.", dataSource);
-        return;
-      }
+    // compute() is used to ensure that the operation for a data source is executed atomically
+    dataSources.compute(
+        dataSource,
+        (dataSourceName, dataSourceState) -> {
+          if (dataSourceState == null) {
+            log.info("Told to delete a queryable for a dataSource[%s] that doesn't exist.", dataSourceName);
+          } else {
+            final VersionedIntervalTimeline<String, ReferenceCountingSegment> loadedIntervals =
+                dataSourceState.getTimeline();
 
-      PartitionChunk<ReferenceCountingSegment> removed = loadedIntervals.remove(
-          segment.getInterval(),
-          segment.getVersion(),
-          segment.getShardSpec().createChunk(null)
-      );
-      ReferenceCountingSegment oldQueryable = (removed == null) ? null : removed.getObject();
+            final PartitionChunk<ReferenceCountingSegment> removed = loadedIntervals.remove(
+                segment.getInterval(),
+                segment.getVersion(),
+                segment.getShardSpec().createChunk(null)
+            );
+            final ReferenceCountingSegment oldQueryable = (removed == null) ? null : removed.getObject();
 
-      if (oldQueryable != null) {
-        synchronized (dataSourceSizes) {
-          dataSourceSizes.add(dataSource, -segment.getSize());
-        }
-        synchronized (dataSourceCounts) {
-          dataSourceCounts.add(dataSource, -1L);
-        }
+            if (oldQueryable != null) {
+              dataSourceState.removeSegment(segment);
```

Comment:
Won't it work to check if there are no more segments in this `DataSourceState`, and then just return `null` from this Map.compute's callback, instead of existing `DataSourceState`?